File: avro/avro-annotation-processor/src/main/java/ru/tinkoff/kora/avro/annotation/processor/reader/AvroReaderGenerator.java
package ru.tinkoff.kora.avro.annotation.processor.reader;

import com.squareup.javapoet.*;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.avro.annotation.processor.AvroTypes;
import ru.tinkoff.kora.avro.annotation.processor.AvroUtils;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.TypeParameterElement;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;

public class AvroReaderGenerator {

    private final ProcessingEnvironment env;

    public AvroReaderGenerator(ProcessingEnvironment processingEnvironment) {
        this.env = processingEnvironment;
    }

    public void generateBinary(TypeElement element) {
        var typeName = TypeName.get(element.asType());
        var typeBuilder = TypeSpec.classBuilder(AvroUtils.readerBinaryName(element))
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", AvroReaderGenerator.class.getCanonicalName()))
                .build())
            .addSuperinterface(ParameterizedTypeName.get(AvroTypes.reader, typeName))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(element);

        for (TypeParameterElement typeParameter : element.getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }

        typeBuilder.addField(FieldSpec.builder(AvroTypes.schema, "SCHEMA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("$T.getClassSchema()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(AvroTypes.specificData, "SPECIFIC_DATA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T().getSpecificData()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(ParameterizedTypeName.get(AvroTypes.datumReader, typeName), "READER")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T<>(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader)
            .build());

        var method = MethodSpec.methodBuilder("read")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addException(IOException.class)
            .addAnnotation(Nullable.class)
            .addAnnotation(Override.class)
            .addParameter(TypeName.get(InputStream.class), "value")
            .returns(typeName);
        method.beginControlFlow("if (value == null || value.available() == 0)");
        method.addStatement("return null");
        method.endControlFlow();
        method.addStatement("var decoder = $T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory);
        method.addStatement("return READER.read(new $T(), decoder)", typeName);

        typeBuilder.addMethod(method.build());
        TypeSpec spec = typeBuilder.build();

        var packageElement = AvroUtils.classPackage(this.env.getElementUtils(), element);
        var javaFile = JavaFile.builder(packageElement, spec).build();
        CommonUtils.safeWriteTo(this.env, javaFile);
    }

    public void generateJson(TypeElement element) {
        var typeName = TypeName.get(element.asType());
        var typeBuilder = TypeSpec.classBuilder(AvroUtils.readerJsonName(element))
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", AvroReaderGenerator.class.getCanonicalName()))
                .build())
            .addSuperinterface(ParameterizedTypeName.get(AvroTypes.reader, typeName))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(element);

        for (TypeParameterElement typeParameter : element.getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }

        typeBuilder.addField(FieldSpec.builder(AvroTypes.schema, "SCHEMA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("$T.getClassSchema()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(AvroTypes.specificData, "SPECIFIC_DATA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T().getSpecificData()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(ParameterizedTypeName.get(AvroTypes.datumReader, typeName), "READER")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T<>(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader)
            .build());

        var method = MethodSpec.methodBuilder("read")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addException(IOException.class)
            .addAnnotation(Nullable.class)
            .addAnnotation(Override.class)
            .addParameter(TypeName.get(InputStream.class), "value")
            .returns(typeName);
        method.beginControlFlow("if (value == null || value.available() == 0)");
        method.addStatement("return null");
        method.endControlFlow();
        method.addStatement("var decoder = $T.get().jsonDecoder(SCHEMA, value)", AvroTypes.decoderFactory);
        method.addStatement("return READER.read(new $T(), decoder)", typeName);

        typeBuilder.addMethod(method.build());
        TypeSpec spec = typeBuilder.build();

        var packageElement = AvroUtils.classPackage(this.env.getElementUtils(), element);
        var javaFile = JavaFile.builder(packageElement, spec).build();
        CommonUtils.safeWriteTo(this.env, javaFile);
    }
}

File: avro/avro-symbol-processor/src/main/kotlin/ru/tinkoff/kora/avro/symbol/processor/reader/AvroReaderGenerator.kt
package ru.tinkoff.kora.avro.symbol.processor.reader

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.jvm.throws
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
import com.squareup.kotlinpoet.ksp.toTypeVariableName
import com.squareup.kotlinpoet.ksp.writeTo
import ru.tinkoff.kora.avro.symbol.processor.AvroTypes
import ru.tinkoff.kora.avro.symbol.processor.classPackage
import ru.tinkoff.kora.avro.symbol.processor.readerBinaryName
import ru.tinkoff.kora.avro.symbol.processor.readerJsonName
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.KspCommonUtils.toTypeName
import java.io.IOException
import java.io.InputStream

class AvroReaderGenerator(val resolver: Resolver, private val codeGenerator: CodeGenerator) {

    fun generateBinary(declaration: KSClassDeclaration) {
        val typeName = declaration.toTypeName()
        val typeParameterResolver = declaration.typeParameters.toTypeParameterResolver()
        val readerInterface = AvroTypes.reader.parameterizedBy(typeName)
        val typeBuilder = TypeSpec.classBuilder(declaration.readerBinaryName())
            .generated(AvroReaderGenerator::class)
        declaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }

        typeBuilder.addSuperinterface(readerInterface)

        declaration.typeParameters.forEach {
            typeBuilder.addTypeVariable(it.toTypeVariableName(typeParameterResolver))
        }

        typeBuilder.addProperty(
            PropertySpec.builder("SCHEMA", AvroTypes.schema)
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T.getClassSchema()", typeName)
                .build()
        )
        typeBuilder.addProperty(
            PropertySpec.builder("SPECIFIC_DATA", AvroTypes.specificData)
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T().getSpecificData()", typeName)
                .build()
        )
        typeBuilder.addProperty(
            PropertySpec.builder("READER", AvroTypes.datumReader.parameterizedBy(typeName))
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader)
                .build()
        )

        val method = FunSpec.builder("read")
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
            .throws(IOException::class)
            .addParameter("value", InputStream::class.asTypeName().copy(true))
            .returns(typeName.copy(true))
        method.beginControlFlow("if (value == null || value.available() == 0)")
        method.addStatement("return null")
        method.endControlFlow()
        method.addStatement("val decoder = %T.get().directBinaryDecoder(value, null)", AvroTypes.decoderFactory)
        method.addStatement("return READER.read(%T(), decoder)", typeName)

        typeBuilder.addFunction(method.build())
        val spec = typeBuilder.build()

        val packageElement = classPackage(declaration)
        val fileSpec = FileSpec.builder(
            packageName = packageElement,
            fileName = spec.name!!
        )
        fileSpec.addType(spec)
        fileSpec.build().writeTo(codeGenerator = codeGenerator, aggregating = false)
    }

    fun generateJson(declaration: KSClassDeclaration) {
        val typeName = declaration.toTypeName()
        val typeParameterResolver = declaration.typeParameters.toTypeParameterResolver()
        val readerInterface = AvroTypes.reader.parameterizedBy(typeName)
        val typeBuilder = TypeSpec.classBuilder(declaration.readerJsonName())
            .generated(AvroReaderGenerator::class)
        declaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }

        typeBuilder.addSuperinterface(readerInterface)

        declaration.typeParameters.forEach {
            typeBuilder.addTypeVariable(it.toTypeVariableName(typeParameterResolver))
        }

        typeBuilder.addProperty(
            PropertySpec.builder("SCHEMA", AvroTypes.schema)
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T.getClassSchema()", typeName)
                .build()
        )
        typeBuilder.addProperty(
            PropertySpec.builder("SPECIFIC_DATA", AvroTypes.specificData)
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T().getSpecificData()", typeName)
                .build()
        )
        typeBuilder.addProperty(
            PropertySpec.builder("READER", AvroTypes.datumReader.parameterizedBy(typeName))
                .addModifiers(KModifier.PRIVATE, KModifier.FINAL)
                .initializer("%T(SCHEMA, SCHEMA, SPECIFIC_DATA)", AvroTypes.datumReader)
                .build()
        )

        val method = FunSpec.builder("read")
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
            .throws(IOException::class)
            .addParameter("value", InputStream::class.asTypeName().copy(true))
            .returns(typeName.copy(true))
        method.beginControlFlow("if (value == null || value.available() == 0)")
        method.addStatement("return null")
        method.endControlFlow()
        method.addStatement("val decoder = %T.get().jsonDecoder(SCHEMA, value)", AvroTypes.decoderFactory)
        method.addStatement("return READER.read(%T(), decoder)", typeName)

        typeBuilder.addFunction(method.build())
        val spec = typeBuilder.build()

        val packageElement = classPackage(declaration)
        val fileSpec = FileSpec.builder(
            packageName = packageElement,
            fileName = spec.name!!
        )
        fileSpec.addType(spec)
        fileSpec.build().writeTo(codeGenerator = codeGenerator, aggregating = false)
    }
}
